### PR TITLE
Remove NotConcrete error

### PR DESCRIPTION
**Pull request by chungquantin**
# Description
- Link to issue: https://github.com/paritytech/polkadot-sdk/issues/3651

polkadot address: 19nSqFQorfF2HxD3oBzWM3oCh4SaCRKWt1yvmgaPYGCo71J

### CODE REVIEW

File: polkadot/xcm/xcm-executor/src/traits/asset_transfer.rs [hunk 1]
```diff
@@ -23,8 +23,6 @@ use xcm::prelude::*;
 /// Errors related to determining asset transfer support.
 #[derive(Copy, Clone, Encode, Decode, Eq, PartialEq, Debug, TypeInfo)]
 pub enum Error {
-	/// Invalid non-concrete asset.
-	NotConcrete,
 	/// Reserve chain could not be determined for assets.
 	UnknownReserve,
```
Comment (ggwpez) [inline polkadot/xcm/xcm-executor/src/traits/asset_transfer.rs:27]: Is this part of the XCM standard @franciscoaguirre?  
Then the following variant would need to get explicit index 1, to not break compatibility.

Comment (chungquantin) [reply polkadot/xcm/xcm-executor/src/traits/asset_transfer.rs:28]: I added the explicit index 1 in this commit: https://github.com/paritytech/polkadot-sdk/pull/3867/commits/37f4e30962fedff3ca4f3b8a1029b1187a098a63

**Review by acatangiu — APPROVED** [review]

File: polkadot/xcm/xcm-executor/src/traits/asset_transfer.rs (hunk 1 above)
Comment (franciscoaguirre) [reply polkadot/xcm/xcm-executor/src/traits/asset_transfer.rs:28]: It's not a part of the XCM standard, it's only used in pallet-xcm for determining whether a teleport or a reserve asset transfer needs to be done. So we can remove this explicit index.
Also, I'd add a comment about this, probably on the module itself.

File: prdoc/pr_3867.prdoc [hunk 2]
```diff
@@ -0,0 +1,9 @@
+title: Remove unused NonConcrete error in asset transfer
```
Comment (franciscoaguirre) [inline prdoc/pr_3867.prdoc:1]: I'd say this PR doesn't need a PRDoc, it's just cleanup. If you add the R0-silent label then CI won't check for a PRDoc. I can add it if you agree.

File: polkadot/xcm/xcm-executor/src/traits/asset_transfer.rs [hunk 3]
```diff
@@ -23,9 +23,8 @@ use xcm::prelude::*;
 /// Errors related to determining asset transfer support.
 #[derive(Copy, Clone, Encode, Decode, Eq, PartialEq, Debug, TypeInfo)]
 pub enum Error {
-	/// Invalid non-concrete asset.
-	NotConcrete,
```
Comment (franciscoaguirre) [inline polkadot/xcm/xcm-executor/src/traits/asset_transfer.rs:27]: This error is still being used in `pallet-xcm`, in a conversion from this error type to the pallet's error type. It has to be removed from there as well.

Comment (franciscoaguirre) [reply polkadot/xcm/xcm-executor/src/traits/asset_transfer.rs:27]: There's also a matching error in `pallet-xcm` called [`InvalidAssetNotConcrete`](https://github.com/paritytech/polkadot-sdk/blob/eb6f5abee64e979dba25924f71ef86d2b3ca2deb/polkadot/xcm/pallet-xcm/src/lib.rs#L539). Should be removed as well.

File: prdoc/pr_3867.prdoc (hunk 2 above)
Comment (chungquantin) [reply prdoc/pr_3867.prdoc]: Thank you for your feedback. I removed the prdoc in this commit: https://github.com/paritytech/polkadot-sdk/pull/3867/commits/d1c65ce6ca25b919eb491bb32c64f442d6260cec

As I don't have a permission, could you help me to label this PR as `R0-silent`. Thanks

File: polkadot/xcm/xcm-executor/src/traits/asset_transfer.rs (hunk 3 above)
Comment (chungquantin) [reply polkadot/xcm/xcm-executor/src/traits/asset_transfer.rs:27]: I removed all the relevant code to `NotConcrete` in this commit: https://github.com/paritytech/polkadot-sdk/pull/3867/commits/d1c65ce6ca25b919eb491bb32c64f442d6260cec

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5694557

File: polkadot/xcm/xcm-executor/src/traits/asset_transfer.rs [hunk 4]
```diff
@@ -23,9 +23,8 @@ use xcm::prelude::*;
...
 	/// Reserve chain could not be determined for assets.
+	#[codec(index = 1)]
```
Comment (franciscoaguirre) [inline polkadot/xcm/xcm-executor/src/traits/asset_transfer.rs:27]: No need to put this

Comment (chungquantin) [reply polkadot/xcm/xcm-executor/src/traits/asset_transfer.rs]: Removed in the commit: https://github.com/paritytech/polkadot-sdk/pull/3867/commits/819dae58dc8946802903bd65800bd6b31c32fc26

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5697792

**Review by franciscoaguirre — APPROVED** [review]
👍 Should be good to merge after CI passes

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5700322

**Comment by chungquantin** [comment]
@franciscoaguirre the integration tests keeps failing at `transaction_recrusion_limit_works()` and I saw it logged stack limit error. Guess it is not related to the changes in this request right? 
```
running 1 test
test transact_recursion_limit_works ... FAILED
failures:
failures:
    transact_recursion_limit_works
test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 4 filtered out; finished in 2.18s
--- TRY 6 STDERR:        xcm-executor-integration-tests transact_recursion_limit_works ---
Mar 30 02:18:31.373 ERROR xcm::pallet_xcm::execute: XCM execution failed with error ExceedsStackLimit    
Mar 30 02:18:31.373 ERROR xcm::pallet_xcm::execute: XCM execution failed with error ExpectationFalse    
Mar 30 02:18:31.373 ERROR xcm::pallet_xcm::execute: XCM execution failed with error ExpectationFalse    
Mar 30 02:18:31.373 ERROR xcm::pallet_xcm::execute: XCM execution failed with error ExpectationFalse    
Mar 30 02:18:31.373 ERROR xcm::pallet_xcm::execute: XCM execution failed with error ExpectationFalse    
Mar 30 02:18:31.373 ERROR xcm::pallet_xcm::execute: XCM execution failed with error ExpectationFalse    
Mar 30 02:18:31.373 ERROR xcm::pallet_xcm::execute: XCM execution failed with error ExpectationFalse    
Mar 30 02:18:31.373 ERROR xcm::pallet_xcm::execute: XCM execution failed with error ExpectationFalse    
Mar 30 02:18:31.373 ERROR xcm::pallet_xcm::execute: XCM execution failed with error ExpectationFalse    
Mar 30 02:18:31.373 ERROR xcm::pallet_xcm::execute: XCM execution failed with error ExpectationFalse    
Mar 30 02:18:31.374 ERROR xcm::pallet_xcm::execute: XCM execution failed with error ExpectationFalse    
Mar 30 02:18:31.375 ERROR xcm::pallet_xcm::execute: XCM execution failed with error ExceedsStackLimit    
Mar 30 02:18:31.375 ERROR xcm::pallet_xcm::execute: XCM execution failed with error ExpectationFalse    
Mar 30 02:18:31.375 ERROR xcm::pallet_xcm::execute: XCM execution failed with error ExpectationFalse    
Mar 30 02:18:31.375 ERROR xcm::pallet_xcm::execute: XCM execution failed with error ExpectationFalse    
Mar 30 02:18:31.375 ERROR xcm::pallet_xcm::execute: XCM execution failed with error ExpectationFalse    
Mar 30 02:18:31.375 ERROR xcm::pallet_xcm::execute: XCM execution failed with error ExpectationFalse    
Mar 30 02:18:31.375 ERROR xcm::pallet_xcm::execute: XCM execution failed with error ExpectationFalse    
Mar 30 02:18:31.375 ERROR xcm::pallet_xcm::execute: XCM execution failed with error ExpectationFalse    
Mar 30 02:18:31.376 ERROR xcm::pallet_xcm::execute: XCM execution failed with error ExpectationFalse    
Mar 30 02:18:31.376 ERROR xcm::pallet_xcm::execute: XCM execution failed with error ExpectationFalse    
Mar 30 02:18:31.376 ERROR xcm::pallet_xcm::execute: XCM execution failed with error ExpectationFalse    
thread 'transact_recursion_limit_works' panicked at polkadot/xcm/xcm-executor/integration-tests/src/lib.rs:146:9:
assertion `left == right` failed
  left: 0
 right: 10
stack backtrace:
   0: rust_begin_unwind
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:645:5
   1: core::panicking::panic_fmt
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:72:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
   4: environmental::using
   5: core::ops::function::FnOnce::call_once
   6: core::ops::function::FnOnce::call_once
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:250:5
```

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5712591

**Comment by acatangiu** [comment]
Try merging master, CI should pass now

**Comment by chungquantin** [comment]
@acatangiu I think this PR is still impacted by the CI

**Comment by ggwpez** [comment]
Yes please fix the issue as pinged by the bot: https://github.com/paritytech/polkadot-sdk/pull/3867#issuecomment-2056560502

**Comment by acatangiu** [comment]
> Yes please fix the issue as pinged by the bot: [#3867 (comment)](https://github.com/paritytech/polkadot-sdk/pull/3867#issuecomment-2056560502)
> 
> PS: Dont know why the error index is hard-coded, no wonder it fails. I would personally just do `Err(crate::Error::<Test>::TooManyReserves.into())` instead of hard-coding error indices...

Yes, good idea!

@chungquantin you can locally test this with `cargo test -p pallet-xcm` - fix the tests to pass locally, then should also pass in CI

**Comment by chungquantin** [comment]
@acatangiu @ggwpez I have updated the code to resolve issues with CI.

**Review by liamaharon — APPROVED** [review]

File: polkadot/xcm/pallet-xcm/src/lib.rs [hunk 5]
```diff
@@ -536,17 +536,23 @@ pub mod pallet {
...
 		TooManyReserves,
 		/// Local XCM execution incomplete.
+		#[codec(index = 24)]
 		LocalExecutionIncomplete,
 		/// Could not decode XCM.
+		#[codec(index = 25)]
 		UnableToDecode,
 		/// XCM encoded length is too large.
 		/// Returned when an XCM encoded length is larger than `MaxXcmEncodedSize`.
+		#[codec(index = 26)]
```
Comment (ggwpez) [inline polkadot/xcm/pallet-xcm/src/lib.rs:555]: Now what i meant

Comment (chungquantin) [reply polkadot/xcm/pallet-xcm/src/lib.rs:555]: Oh I saw your suggestion. Should I revert and follow your aproach?

Comment (acatangiu) [reply polkadot/xcm/pallet-xcm/src/lib.rs:555]: no revert required, you could do a followup PR

Comment (chungquantin) [reply polkadot/xcm/pallet-xcm/src/lib.rs:555]: Gotcha thanks @acatangiu @ggwpez

**Pull request merged**